mmdet/apis/ote/extension/utils/hooks.py

```diff
@@ -0,0 +1,253 @@
+    :param min_delta: Minimal decay applied to lr. If the difference between new and old lr is smaller than eps,
+                      the update is ignored
+    """
+    rule_map = {'greater': lambda x, y: x > y, 'less': lambda x, y: x < y}
+    init_value_map = {'greater': -inf, 'less': inf}
+    greater_keys = ['mAP', 'acc', 'top', 'AR@', 'auc', 'precision']
+    less_keys = ['loss']
+
+    def __init__(self, interval: int, metric: str = 'mAP', rule: str = None, patience: int = 3, min_delta: float = 0.0):
+        super().__init__()
+        self.patience = patience
+        self.interval = interval
+        self.min_delta = min_delta
+        self._init_rule(rule, metric)
+
+        self.min_delta *= 1 if self.rule == 'greater' else -1
+        self.best_score = self.init_value_map[self.rule]
+        _, world_size = get_dist_info()
+        self.distributed = True if world_size > 1 else False
+
+    def _init_rule(self, rule, key_indicator):
+        """Initialize rule, key_indicator, comparison_func, and best score.
+        Here is the rule to determine which rule is used for key indicator
+        when the rule is not specific:
+        1. If the key indicator is in ``self.greater_keys``, the rule will be
+           specified as 'greater'.
+        2. Or if the key indicator is in ``self.less_keys``, the rule will be
+           specified as 'less'.
+        3. Or if the key indicator is equal to the substring in any one item
+           in ``self.greater_keys``, the rule will be specified as 'greater'.
+        4. Or if the key indicator is equal to the substring in any one item
+           in ``self.less_keys``, the rule will be specified as 'less'.
+        Args:
+            rule (str | None): Comparison rule for best score.
+            key_indicator (str | None): Key indicator to determine the
+                comparison rule.
+        """
+        if rule not in self.rule_map and rule is not None:
+            raise KeyError(f'rule must be greater, less or None, '
+                           f'but got {rule}.')
+
+        if rule is None:
+            if key_indicator != 'auto':
+                if key_indicator in self.greater_keys:
+                    rule = 'greater'
+                elif key_indicator in self.less_keys:
+                    rule = 'less'
+                elif any(key in key_indicator for key in self.greater_keys):
+                    rule = 'greater'
+                elif any(key in key_indicator for key in self.less_keys):
+                    rule = 'less'
+                else:
+                    raise ValueError(f'Cannot infer the rule for key '
+                                     f'{key_indicator}, thus a specific rule '
+                                     f'must be specified.')
+        self.rule = rule
+        self.key_indicator = key_indicator
+        if self.rule is not None:
+            self.compare_func = self.rule_map[self.rule]
```

Comment:
Is it Ok to have `self.rule` equal to `None` at this point? If no, the error should be raised.